// @sanity/runtime-cli/dist/utils/child-process-wrapper.js
import {table as renderTable} from 'node:console'
import {existsSync, statSync} from 'node:fs'
import {isAbsolute, join} from 'node:path'
import process from 'node:process'
import {pathToFileURL} from 'node:url'
import {inspect} from 'node:util'

export function getFunctionSource(src) {
  const pathToCheck = isAbsolute(src) ? src : join(process.cwd(), src)

  if (statSync(pathToCheck).isDirectory()) {
    const indexPath = join(pathToCheck, 'index.js')
    if (!existsSync(indexPath)) {
      throw Error(`Function directory ${pathToCheck} has no index.js`)
    }
    return pathToFileURL(indexPath).href
  }
  return pathToFileURL(pathToCheck).href
}

// Monkey patch console menthods to have logs match server log format
function logPrefix(level, ...args) {
  const date = new Date()
  const shouldUseColors = process.env.FORCE_COLOR === '1'
  const message = args
    .map((arg) =>
      typeof arg === 'string' ? arg : inspect(arg, {depth: null, colors: shouldUseColors}),
    )
    .join(' ')
  return `${date.toLocaleDateString()} ${date.toLocaleTimeString()} ${level.toUpperCase()} ${message}`
}

console.log = (...args) => {
  process.stdout.write(`${logPrefix('info', ...args)}\n`)
}

console.info = (...args) => {
  process.stdout.write(`${logPrefix('info', ...args)}\n`)
}

console.dir = (obj, options) => {
  const shouldUseColors = process.env.FORCE_COLOR === '1'
  const inspectOptions = {...options, colors: shouldUseColors}
  process.stdout.write(`${logPrefix('info', inspect(obj, inspectOptions))}\n`)
}

console.table = (data, columns) => {
  const prefix = logPrefix('info')

  // Capture table output
  const originalWrite = process.stdout.write.bind(process.stdout)
  let buffer = ''

  process.stdout.write = (chunk, _encoding, callback) => {
    buffer += chunk
    if (typeof callback === 'function') callback()
  }

  renderTable(data, columns)

  // Restore and print
  process.stdout.write = originalWrite
  for (const line of buffer.split('\n')) {
    if (line.trim()) {
      const message = `${!line.startsWith(prefix) ? `${prefix} ${line}` : line}`.trim()
      process.stdout.write(`${message}\n`)
    }
  }
}

console.warn = (...args) => {
  process.stdout.write(`${logPrefix('warn', ...args)}\n`)
}

console.error = (...args) => {
  process.stdout.write(`${logPrefix('error', ...args)}\n`)
}

// Start when payload data arrives from parent process
process.on('message', async (data) => {
  let jsonData = null
  try {
    jsonData = JSON.parse(data)
  } catch {
    // invalid payload so return early
    return
  }

  const {srcPath, payload} = jsonData
  const {context, ...event} = payload

  let logs = ''
  let errorLogs = ''
  let json = null

  // Capture stdout and stderr
  const originalStdoutWrite = process.stdout.write.bind(process.stdout)
  const originalStderrWrite = process.stderr.write.bind(process.stderr)

  process.stdout.write = (chunk) => {
    if (typeof chunk === 'string') logs += chunk
    return true
  }
  process.stderr.write = (chunk) => {
    if (typeof chunk === 'string') errorLogs += chunk
    return true
  }

  try {
    // Import the function code
    const entry = await import(getFunctionSource(srcPath))
    const eventHandler = entry.handler || entry.default

    if (typeof eventHandler !== 'function') {
      throw new Error(
        'No valid handler found. Please provide a default export or a named export, "handler"',
      )
    }

    json = await eventHandler({context, event})

    // Restore streams
    process.stdout.write = originalStdoutWrite
    process.stderr.write = originalStderrWrite

    // Send result to parent
    process.send(JSON.stringify({json, logs, errorLogs}))
  } catch (err) {
    // Restore streams on error
    process.stdout.write = originalStdoutWrite
    process.stderr.write = originalStderrWrite

    const errorInfo = {
      json: null,
      logs,
      errorLogs,
      error: {
        message: err?.message,
        stack: err?.stack,
        name: err?.name,
      },
    }
    process.send(JSON.stringify(errorInfo))
    process.exit(1)
  }
})
