// @sanity/runtime-cli/dist/utils/logger.js
import { format } from 'node:util';
import ora from 'ora';
var LogLevel;
(function (LogLevel) {
    LogLevel[LogLevel["TRACE"] = 0] = "TRACE";
    LogLevel[LogLevel["VERBOSE"] = 1] = "VERBOSE";
    LogLevel[LogLevel["INFO"] = 2] = "INFO";
    LogLevel[LogLevel["WARN"] = 3] = "WARN";
    LogLevel[LogLevel["ERROR"] = 4] = "ERROR";
})(LogLevel || (LogLevel = {}));
export function Logger(log, flags = {}) {
    const logger = (msg) => {
        log(msg);
    };
    const level = flags.verbose ? LogLevel.VERBOSE : flags.trace ? LogLevel.TRACE : LogLevel.INFO;
    logger.trace = (formatter, ...args) => level <= LogLevel.TRACE && logger(format(formatter, ...args));
    logger.verbose = (formatter, ...args) => level <= LogLevel.VERBOSE && logger(format(formatter, ...args));
    logger.info = (formatter, ...args) => level <= LogLevel.INFO && logger(format(formatter, ...args));
    logger.warn = (formatter, ...args) => level <= LogLevel.WARN && logger(format(formatter, ...args));
    logger.error = (formatter, ...args) => level <= LogLevel.ERROR && logger(format(formatter, ...args));
    const oraWrapper = (opts) => {
        if (level >= LogLevel.INFO)
            return ora(opts);
        return createOraLineLoggingWrapper(opts, logger);
    };
    logger.ora = oraWrapper;
    return logger;
}
/**
 * An ora-like wrapper that uses standard line output logging instead of rendering a spinner; used in verbose and trace modes.
 */
function createOraLineLoggingWrapper(opts, logger) {
    const text = typeof opts === 'string' ? opts : (typeof opts === 'object' ? opts.text : '') || '';
    const wrapper = {
        clear() {
            logger('');
            return wrapper;
        },
        fail(subtext) {
            logger.error(subtext);
            return wrapper;
        },
        info(subtext) {
            logger(subtext || wrapper.text);
            return wrapper;
        },
        start(subtext) {
            logger(subtext || wrapper.text);
            return wrapper;
        },
        stop() {
            return wrapper;
        },
        succeed(subtext) {
            logger(subtext || wrapper.text);
            return wrapper;
        },
        text,
    };
    return wrapper;
}
