// @sanity/migrate/dist/fetch-utils/sanityRequestOptions.js
import { createRequire } from 'node:module';
const require = createRequire(import.meta.url);
function getUserAgent() {
    if (globalThis.window === undefined) {
        // only set UA if we're in a non-browser environment
        try {
            const pkg = require('../../package.json');
            return `${pkg.name}@${pkg.version}`;
            // eslint-disable-next-line no-empty
        }
        catch { }
    }
    return null;
}
function normalizeApiHost(apiHost) {
    return apiHost.replace(/^https?:\/\//, '');
}
export function toFetchOptions(req) {
    const { apiHost, apiVersion, body, endpoint, projectId, tag, token } = req;
    const requestInit = {
        headers: {
            'Content-Type': 'application/json',
        },
        method: endpoint.method || 'GET',
        ...(body !== undefined && { body }),
    };
    const ua = getUserAgent();
    if (ua) {
        requestInit.headers = {
            ...requestInit.headers,
            'User-Agent': ua,
        };
    }
    if (token) {
        requestInit.headers = {
            ...requestInit.headers,
            Authorization: `bearer ${token}`,
        };
    }
    const normalizedApiHost = normalizeApiHost(apiHost);
    const path = `/${apiVersion}${endpoint.path}`;
    const host = endpoint.global ? normalizedApiHost : `${projectId}.${normalizedApiHost}`;
    const searchParams = new URLSearchParams([
        ...endpoint.searchParams,
        ...(tag ? [['tag', tag]] : []),
    ]).toString();
    return {
        init: requestInit,
        url: `https://${host}/${path}${searchParams ? `?${searchParams}` : ''}`,
    };
}
