// @sanity/runtime-cli/dist/commands/functions/env/add.js
import { Args } from '@oclif/core';
import { DeployedStackCommand } from '../../../baseCommands.js';
import { functionEnvAddCore } from '../../../cores/functions/env/add.js';
import { Logger } from '../../../utils/logger.js';
export default class EnvAddCommand extends DeployedStackCommand {
    static summary = 'Add or set an environment variable for a deployed function';
    static description = `Sets an environment variable in a deployed Sanity Function. If the variable already exists, its value is updated.

Environment variables are useful for API keys, configuration values, and other secrets that shouldn't be hardcoded. Changes take effect on the next function invocation.`;
    static args = {
        name: Args.string({ description: 'The name of the Sanity Function', required: true }),
        key: Args.string({ description: 'The name of the environment variable', required: true }),
        value: Args.string({ description: 'The value of the environment variable', required: true }),
    };
    static examples = [
        '<%= config.bin %> <%= command.id %> MyFunction API_URL https://api.example.com/',
    ];
    async run() {
        const { success, error } = await functionEnvAddCore({
            bin: this.config.bin,
            log: Logger(this.log.bind(this), this.flags),
            args: this.args,
            auth: this.auth,
            blueprint: this.blueprint,
            deployedStack: this.deployedStack,
            scopeType: this.scopeType,
            scopeId: this.scopeId,
            token: this.sanityToken,
            stackId: this.stackId,
            validateResources: this.flags['validate-resources'],
        });
        if (!success)
            this.error(error);
    }
}
