// @sanity/runtime-cli/dist/commands/functions/env/remove.js
import { Args } from '@oclif/core';
import { DeployedStackCommand } from '../../../baseCommands.js';
import { functionEnvRemoveCore } from '../../../cores/functions/env/remove.js';
import { Logger } from '../../../utils/logger.js';
export default class EnvRemoveCommand extends DeployedStackCommand {
    static summary = 'Remove an environment variable from a deployed function';
    static description = `Deletes an environment variable from a deployed Sanity Function. The change takes effect on the next function invocation.

Use 'functions env list' to see current variables before removing.`;
    static args = {
        name: Args.string({ description: 'The name of the Sanity Function', required: true }),
        key: Args.string({ description: 'The name of the environment variable', required: true }),
    };
    static examples = ['<%= config.bin %> <%= command.id %> MyFunction API_URL'];
    async run() {
        const { success, error } = await functionEnvRemoveCore({
            bin: this.config.bin,
            log: Logger(this.log.bind(this), this.flags),
            args: this.args,
            auth: this.auth,
            blueprint: this.blueprint,
            deployedStack: this.deployedStack,
            scopeType: this.scopeType,
            scopeId: this.scopeId,
            token: this.sanityToken,
            stackId: this.stackId,
            validateResources: this.flags['validate-resources'],
        });
        if (!success)
            this.error(error);
    }
}
