// @sanity/blueprints/src/definers/roles.ts
import {
  type BlueprintProjectRoleResource,
  type BlueprintRoleConfig,
  type BlueprintRoleResource,
  validateProjectRole,
  validateRole,
} from '../index.js'
import {runValidation} from '../utils/validation.js'

/*
 * FUTURE example (move below @example when ready)
 * @example All options
 * ```ts
 * defineRole({
 *   name: 'ci-deploy-role',
 *   title: 'CI Deploy Role',
 *   description: 'Permissions for CI/CD pipelines to manage CORS and datasets',
 *   appliesToRobots: true,
 *   appliesToUsers: false,
 *   permissions: [{
 *     name: 'sanity-project-cors',
 *     action: 'create',
 *   }, {
 *     name: 'sanity-project-cors',
 *     action: 'read',
 *   }, {
 *     name: 'sanity-project-cors',
 *     action: 'delete',
 *   }],
 * })
 * ```
 */
/**
 * Defines a role that is scoped to the same resource as the blueprint.
 *
 * ```ts
 * defineRole({
 *   name: 'custom-robot-role',
 *   title: 'Custom Robot Role',
 *   appliesToRobots: true,
 *   permissions: [{
 *     name: 'sanity-project-cors',
 *     action: 'create',
 *   }],
 * })
 * ```
 * @param parameters The configuration of the role
 * @public
 * @beta Deploying Roles via Blueprints is experimental. This feature is stabilizing but may still be subject to breaking changes.
 * @category Definers
 * @expandType BlueprintRoleConfig
 * @returns The role resource
 */
export function defineRole(parameters: BlueprintRoleConfig, options?: {skipValidation?: boolean}): BlueprintRoleResource {
  const roleResource: BlueprintRoleResource = {
    name: parameters.name,
    type: 'sanity.access.role',
    title: parameters.title,
    description: parameters.description,
    appliesToUsers: parameters.appliesToUsers,
    appliesToRobots: parameters.appliesToRobots,
    permissions: parameters.permissions,
  }

  if (parameters.lifecycle) {
    roleResource.lifecycle = parameters.lifecycle
  }

  if (options?.skipValidation !== true) runValidation(() => validateRole(roleResource))

  return roleResource
}

/*
 * FUTURE example (move below @example when ready)
 * @example Cross-resource references
 * ```ts
 * const role = defineProjectRole(projectId, {
 *   name: 'ci-deploy-role',
 *   title: 'CI Deploy Role',
 *   appliesToRobots: true,
 *   permissions: [{
 *     name: 'sanity-project-cors',
 *     action: 'create',
 *   }, {
 *     name: 'sanity-project-cors',
 *     action: 'delete',
 *   }],
 * })
 *
 * defineRobotToken({
 *   name: 'ci-robot',
 *   memberships: [{
 *     resourceType: 'project',
 *     resourceId: projectId,
 *     roleNames: ['$.resources.ci-deploy-role'],
 *   }],
 * })
 * ```
 */
/**
 * Defines a role that is scoped to the specified project.
 *
 * ```ts
 * defineProjectRole(projectId, {
 *   name: 'viewer-role',
 *   title: 'Viewer',
 *   appliesToUsers: true,
 *   permissions: [{
 *     name: 'sanity-project-cors',
 *     action: 'read',
 *   }],
 * })
 * ```
 * @public
 * @beta Deploying Roles via Blueprints is experimental. This feature is stabilizing but may still be subject to breaking changes.
 * @hidden
 * @category Definers
 * @expandType BlueprintRoleConfig
 * @param projectId The ID of the project to which the role will be scoped
 * @param parameters The configuration of the role
 * @returns The role resource
 */
export function defineProjectRole(projectId: string, parameters: BlueprintRoleConfig): BlueprintProjectRoleResource {
  const roleResource = defineRole(parameters, {skipValidation: true})

  const projectRoleResource: BlueprintProjectRoleResource = {
    ...roleResource,
    resourceType: 'project',
    resourceId: projectId,
  }

  runValidation(() => validateProjectRole(projectRoleResource))

  return projectRoleResource
}
