// @sanity/runtime-cli/dist/utils/functions/detect-native-modules.js
import { existsSync, readdirSync, readFileSync, statSync } from 'node:fs';
import path from 'node:path';
import { cwd } from 'node:process';
import { createPnpmRequire } from '../pnpm.js';
/**
 * Patterns to identify native modules based on common file types and build configurations.
 */
const KNOWN_MODULE_PATTERNS = [/binding\.gyp$/, /\.node$/];
/**
 * Recurse node_modules for native modules.
 * Native modules built on one platform may not work on another (e.g., macOS vs Linux)
 */
export const detectNativeModules = (zip) => {
    try {
        const zippedEntries = zip.getEntries();
        const moduleNames = zippedEntries
            // Filter entries that match known native module patterns
            .filter((entry) => KNOWN_MODULE_PATTERNS.some((pattern) => pattern.test(entry.entryName)))
            // Extract package names from the entry paths
            .map((entry) => entry.entryName.match(/node_modules[\\/](.+?)[\\/]/)?.[1])
            // Filter out undefined results (in case the regex didn't match)
            .filter((name) => name !== undefined);
        // return a unique list of module names
        return [...new Set(moduleNames)];
    }
    catch (error) {
        const message = error instanceof Error ? error.message : String(error);
        throw new Error(`Failed to scan zip for native modules: ${message}`);
    }
};
/**
 * Recursively checks a directory for files that match known native module patterns.
 */
const hasNativeFiles = (directory) => {
    for (const name of readdirSync(directory)) {
        // don't recurse back into node_modules.
        if (name === 'node_modules')
            continue;
        const fullPath = path.join(directory, name);
        if (statSync(fullPath).isDirectory()) {
            if (hasNativeFiles(fullPath))
                return true;
        }
        else if (KNOWN_MODULE_PATTERNS.some((pattern) => pattern.test(name))) {
            return true;
        }
    }
    return false;
};
/**
 * Finds the dependencies of the function and compares to the workspace root to see if any native modules are included.
 * Due to how PNPM with workspaces symlinks dependencies, we have to check the workspace root for native modules prior to the bundle step.
 */
export const detectNativeModulesForBundle = (resource) => {
    if (!resource.src)
        throw new Error('Resource src is required');
    const hasWorkspace = existsSync(path.join(cwd(), 'pnpm-workspace.yaml'));
    if (!hasWorkspace) {
        return [];
    }
    const sourcePath = path.resolve(cwd(), resource.src);
    const stats = statSync(sourcePath);
    const entryDir = stats.isFile() ? path.dirname(sourcePath) : sourcePath;
    const hasPackageJson = existsSync(path.join(entryDir, 'package.json'));
    if (!hasPackageJson) {
        throw new Error('pnpm workspace detected but no package.json found in function src');
    }
    // gets package.json dependencies of the function. We use these function deps to compare against the workspace root modules.
    const pkgJsonPath = path.join(entryDir, 'package.json');
    // a function should always have deps, but this will just be a little safer.
    const functionDeps = JSON.parse(readFileSync(pkgJsonPath, 'utf-8')).dependencies ?? {};
    const workspaceModules = createPnpmRequire(cwd());
    const nativeModules = [];
    for (const depName of Object.keys(functionDeps)) {
        try {
            const resolved = workspaceModules.resolve(depName);
            let currentDir = path.dirname(resolved);
            while (currentDir !== cwd() && currentDir !== path.parse(currentDir).root) {
                // look for node_module package.json to locate top level
                if (existsSync(path.join(currentDir, 'package.json'))) {
                    if (hasNativeFiles(currentDir)) {
                        nativeModules.push(depName);
                        break;
                    }
                }
                currentDir = path.dirname(currentDir);
            }
        }
        catch (_e) {
            // If we fail to resolve, just skip
        }
    }
    return nativeModules;
};
