// @sanity/runtime-cli/dist/utils/functions/resolve-dependencies.js
import { existsSync } from 'node:fs';
import { join } from 'node:path';
import hydrate from '@architect/hydrate';
import inventory from '@architect/inventory';
import { convertResourceToArcFormat } from './resource-to-arc.js';
export async function resolveResourceDependencies(resource, { transpiled, installer, }) {
    const rawArc = await convertResourceToArcFormat(resource, transpiled);
    const inv = await inventory({ rawArc });
    const cwd = inv.inv._project.cwd;
    let installType = 'npm';
    if (installer) {
        installType = installer;
    }
    else if (existsSync(join(cwd, 'pnpm-lock.yaml'))) {
        installType = 'pnpm';
    }
    else if (existsSync(join(cwd, 'yarn.lock'))) {
        installType = 'yarn';
    }
    try {
        await hydrate.install(toInstallerOptions(inv, installType));
    }
    catch (err) {
        // This is a temporary fix.
        const regex = /ENOTDIR: not a directory, unlink ['"].*[/\\]node_modules['"]/;
        const errorMessage = err instanceof Error ? err.message : String(err);
        if (!regex.test(errorMessage)) {
            throw err;
        }
    }
}
function toInstallerOptions(inv, type) {
    return {
        inventory: inv,
        hydrateShared: false,
        quiet: true,
        npm: type === 'npm',
        pnpm: type === 'pnpm',
        yarn: type === 'yarn',
    };
}
