// @sanity/runtime-cli/dist/commands/functions/dev.js
import { Flags } from '@oclif/core';
import { LocalBlueprintCommand } from '../../baseCommands.js';
import { functionDevCore } from '../../cores/functions/dev.js';
import { Logger } from '../../utils/logger.js';
export default class DevCommand extends LocalBlueprintCommand {
    static summary = 'Start the Sanity Function emulator';
    static description = `Runs a local, web-based development server to test your functions before deploying.

Open the emulator in your browser to interactively test your functions with the payload editor.

Optionally, set the host and port with the --host and --port flags. Function timeout can be configured with the --timeout flag.

To invoke a function with the CLI, use 'functions test'.`;
    static examples = [
        '<%= config.bin %> <%= command.id %> --host 127.0.0.1 --port 8974',
        '<%= config.bin %> <%= command.id %> --timeout 60',
    ];
    static flags = {
        host: Flags.string({
            char: 'h',
            description: 'The local network interface at which to listen. [default: "localhost"]',
            required: false,
        }),
        port: Flags.integer({
            char: 'p',
            description: 'TCP port to start emulator on. [default: 8080]',
            required: false,
        }),
        timeout: Flags.integer({
            char: 't',
            description: 'Maximum execution time for all functions, in seconds. Takes precedence over function-specific `timeout`',
            required: false,
        }),
    };
    async run() {
        const { flags } = await this.parse(DevCommand);
        const { success, error, streaming } = await functionDevCore({
            bin: this.config.bin,
            log: Logger(this.log.bind(this), this.flags),
            validateResources: this.flags['validate-resources'],
            flags,
        });
        if (!success)
            this.error(error);
        if (streaming)
            await streaming;
    }
}
