// @sanity/runtime-cli/dist/config.js
import { readFileSync } from 'node:fs';
import { env } from 'node:process';
import { fileURLToPath } from 'node:url';
import * as pkg from 'empathic/package';
import getToken from './utils/get-token.js';
export const BLUEPRINT_CONFIG_VERSION = 'v2025-05-08';
export const BLUEPRINT_CONFIG_DIR = '.sanity';
export const BLUEPRINT_CONFIG_FILE = 'blueprint.config.json';
export let RUNTIME_CLI_VERSION;
try {
    const packageJsonPath = pkg.up({ cwd: fileURLToPath(new URL('.', import.meta.url)) });
    if (packageJsonPath) {
        const packageJson = JSON.parse(readFileSync(packageJsonPath, 'utf8'));
        if (packageJson.name === '@sanity/runtime-cli') {
            RUNTIME_CLI_VERSION = packageJson.version;
        }
    }
}
catch { }
const nodeEnv = env.NODE_ENV?.toLowerCase() ?? 'production';
const isTest = nodeEnv === 'test';
const sanityEnv = env.SANITY_INTERNAL_ENV?.toLowerCase() ?? 'production';
const sanityProd = sanityEnv === 'production';
const isLive = sanityProd || sanityEnv === 'staging';
const isPublishing = env?.PUBLISHING?.toLowerCase() === 'true';
const apiUrls = {
    production: 'https://api.sanity.io/',
    staging: 'https://api.sanity.work/',
    default: 'http://api.sanity/',
};
const apiUrl = sanityEnv ? (apiUrls[sanityEnv] ?? sanityEnv) : apiUrls.default;
let _token;
function loadToken() {
    let token = null;
    if (isTest || isPublishing) {
        token = 'token';
    }
    else {
        try {
            token = getToken({ prod: sanityProd });
        }
        catch (error) {
            if (error instanceof Error && 'code' in error && error.code === 'ENOENT') {
                // do nothing, validateToken and commands handle missing config file
            }
            else {
                // rethrow other errors like JSON.parse failure
                throw error;
            }
        }
    }
    return token;
}
export default {
    isTest,
    isLive,
    apiUrl,
    get token() {
        return _token === undefined ? loadToken() : _token;
    },
    set token(value) {
        _token = value;
    },
};
