// @sanity/runtime-cli/dist/constants.js
export const SANITY_ACCESS_ROLE = 'sanity.access.role';
export const SANITY_ACCESS_ROBOT = 'sanity.access.robot';
export const SANITY_FUNCTION_PREFIX = 'sanity.function.';
export const SANITY_FUNCTION_DOCUMENT = 'sanity.function.document';
export const SANITY_FUNCTION_MEDIA_LIBRARY_ASSET = 'sanity.function.media-library.asset';
export const SANITY_FUNCTION_SCHEDULED = 'sanity.function.cron';
export const SANITY_PROJECT = 'sanity.project';
export const SANITY_PROJECT_CORS = 'sanity.project.cors';
export const SANITY_PROJECT_DATASET = 'sanity.project.dataset';
export const SANITY_PROJECT_WEBHOOK = 'sanity.project.webhook';
export const EVENT_DOCUMENT_PUBLISH = 'document-publish';
export const EVENT_DOCUMENT_CREATE = 'document-create';
export const EVENT_DOCUMENT_UPDATE = 'document-update';
export const EVENT_DOCUMENT_DELETE = 'document-delete';
export const EVENT_MEDIA_LIBRARY_ASSET_CREATE = 'media-library-asset-create';
export const EVENT_MEDIA_LIBRARY_ASSET_UPDATE = 'media-library-asset-update';
export const EVENT_MEDIA_LIBRARY_ASSET_DELETE = 'media-library-asset-delete';
export const EVENT_SCHEDULED = 'scheduled-function';
export const FUNCTION_TYPES = [
    EVENT_DOCUMENT_PUBLISH,
    EVENT_DOCUMENT_CREATE,
    EVENT_DOCUMENT_UPDATE,
    EVENT_DOCUMENT_DELETE,
    EVENT_MEDIA_LIBRARY_ASSET_CREATE,
    EVENT_MEDIA_LIBRARY_ASSET_UPDATE,
    EVENT_MEDIA_LIBRARY_ASSET_DELETE,
    EVENT_SCHEDULED,
];
export const MAX_ASSET_SIZE = 209_715_200; // 200 MB in bytes
export const CONVERT_BYTES_TO_MB = 1_048_576; // Used to convert bytes to megabytes (1024 * 1024)
export const LABEL_DOCUMENT_CREATE = 'Document Create';
export const LABEL_DOCUMENT_UPDATE = 'Document Update';
export const LABEL_DOCUMENT_DELETE = 'Document Delete';
export const LABEL_MEDIA_LIBRARY_ASSET_CREATE = 'Media Library Asset Create';
export const LABEL_MEDIA_LIBRARY_ASSET_UPDATE = 'Media Library Asset Update';
export const LABEL_MEDIA_LIBRARY_ASSET_DELETE = 'Media Library Asset Delete';
export const LABEL_SCHEDULED = 'Scheduled';
